pkg/charm/instance.go

```diff
@@ -83,6 +40,100 @@ func (i *instance) options(vflag bool) []string {
 			line = fmt.Sprintf("%s (default \"%s\")", line, f.DefValue)
 		}
 		body = append(body, line)
+	}
+	sort.Slice(body, func(i, j int) bool {
+		return strings.ToLower(body[i]) < strings.ToLower(body[j])
 	})
 	return body
 }
+
+func parse(spec *Spec, args []string, parent Command) (Path, []string, bool, error) {
+	var path Path
+	var help, hidden, usage bool
+	flags := flag.NewFlagSet(spec.Name, flag.ContinueOnError)
+	flags.BoolVar(&help, HelpFlag, false, "display help")
+	flags.BoolVar(&hidden, HiddenFlag, false, "show hidden options")
+	flags.Usage = func() {
+		usage = true
+	}
+	for {
+		cmd, err := spec.New(parent, flags)
+		if err != nil {
+			return nil, nil, false, err
+		}
+		component := &instance{
+			spec:    spec,
+			command: cmd,
+		}
+		path = append(path, component)
+		parent = cmd
+		if err := flags.Parse(args); err != nil {
+			if usage {
+				s := strings.Join(args, " ")
+				err = fmt.Errorf("at flag: %q: %w", s, err)
+			}
+			return path, nil, false, err
+		}
+		if help {
+			return path, nil, hidden, NeedHelp
+		}
+		rest := flags.Args()
+		if len(rest) != 0 {
+			spec = component.spec.lookupSub(rest[0])
+			if spec != nil {
+				// We found a subcommand, so continue building the chain.
+				args = rest[1:]
+				continue
+			}
+		}
+		return path, rest, false, nil
+	}
+}
+
+func diff(flags *flag.FlagSet, all map[string]*flag.Flag) map[string]*flag.Flag {
+	difference := make(map[string]*flag.Flag)
+	flags.VisitAll(func(f *flag.Flag) {
+		if _, ok := all[f.Name]; !ok {
+			all[f.Name] = f
+			difference[f.Name] = f
+		}
+	})
+	return difference
+}
+
+func parseHelp(spec *Spec, args []string) (Path, error) {
+	var path Path
+	flags := flag.NewFlagSet(spec.Name, flag.ContinueOnError)
+	var b bool
+	flags.BoolVar(&b, HelpFlag, false, "display help")
+	flags.BoolVar(&b, HiddenFlag, false, "show hidden options")
```

Comment:
Nit: Using `BoolVar` for these results in help output like this:
```
    -h display help (default "false")
    -hidden show hidden options (default "false")
```
Switch to `Func` for this:
```
    -h display help
    -hidden show hidden options
```
```suggestion
	flags.Func(HelpFlag, "display help", func(_ string) error { return nil })
	flags.Func(HiddenFlag, "show hidden options", func(_ string) error { return nil })
```

Reply:
That doesn't seem to work.